relentless/core/variable.py

```diff
@@ -0,0 +1,550 @@
+
+    Parameters
+    ----------
+    a : :py:class:`Variable`
+        The first variable object on which the BinaryOperator depends.
+    b : :py:class:`Variable`
+        The second variable object on which the BinaryOperator depends.
+
+    """
+    def __init__(self, a, b):
+        super().__init__(a=a, b=b)
+
+class ArithmeticMean(BinaryOperator):
+    r"""Binary operator based on arithmetic mean.
+
+    The ArithmeticMean object akes the arithmetic mean of two :py:class:`Variable`s
+    as shown below:
+
+    .. math::
+
+        v = \frac{1}{2}(a+b)
+
+    This class can also be used to implement mixing rules.
+
+    Parameters
+    ----------
+    a : :py:class:`Variable`
+        The first parameter of the ArithmeticMean.
+    b : :py:class:`Variable`
+        The second parameter of the ArithmeticMean.
+
+    """
+    @property
+    def value(self):
+        return 0.5*(self.a.value+self.b.value)
+
+    def _derivative(self, param):
+        """Calculates the derivative of the specified ArithmeticMean object with respect to its parameters.
+
+        Parameters
+        ----------
+        param : str
+            The parameter with respect to which to take the derivative.
+            (Can only be 'a' or 'b').
+
+        Returns
+        -------
+        float
+            The calculated derivative value.
+
+        Raises
+        ------
+        ValueError
+            If the parameter argument is not 'a' or 'b'.
+
+        """
+        if param == 'a':
+            return 0.5
+        elif param == 'b':
+            return 0.5
+        else:
+            raise ValueError('Unknown parameter')
+
+class GeometricMean(BinaryOperator):
+    r"""Binary operator based on geometric mean.
+
+    The GeometricMean object akes the geometric mean of two :py:class:`Variable`s
+    as shown below:
+
+    .. math::
+
+        v = \sqrt{ab}
+
+    This class can also be used to implement mixing rules.
+
+    Parameters
+    ----------
+    a : :py:class:`Variable`
+        The first parameter of the GeometricMean.
+    b : :py:class:`Variable`
+        The second parameter of the GeometricMean.
+
+    """
+    @property
+    def value(self):
+        return np.sqrt(self.a.value*self.b.value)
+
+    def _derivative(self, param):
+        """Calculates the derivative of the specified GeometricMean object with respect to its parameters.
+
+        Parameters
+        ----------
+        param : str
+            The parameter with respect to which to take the derivative.
+            (Can only be 'a' or 'b').
+
+        Returns
+        -------
+        float
+            The calculated derivative value.
+
+        Raises
+        ------
+        ValueError
+            If the parameter argument is not 'a' or 'b'.
+
+        """
+        if param == 'a':
+            return 0.5*np.sqrt(self.b.value/self.a.value)
+        elif param == 'b':
+            return 0.5*np.sqrt(self.a.value/self.b.value)
```

Comment:
Should we add a special case to each of these here for when the denominator is zero? It should return `np.inf` in those cases and not do a divide by zero.